pkg/controller/kfservice/reconcilers/knative/service_reconciler.go

```diff
@@ -0,0 +1,182 @@
+limitations under the License.
+*/
+
+package knative
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha2"
+	"github.com/kubeflow/kfserving/pkg/constants"
+	"github.com/kubeflow/kfserving/pkg/controller/kfservice/resources/knative"
+	"knative.dev/pkg/kmp"
+	knservingv1alpha1 "knative.dev/serving/pkg/apis/serving/v1alpha1"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("Reconciler")
+
+type ServiceReconciler struct {
+	client         client.Client
+	scheme         *runtime.Scheme
+	serviceBuilder *knative.ServiceBuilder
+}
+
+func NewServiceReconciler(client client.Client, scheme *runtime.Scheme, config *v1.ConfigMap) *ServiceReconciler {
+	return &ServiceReconciler{
+		client:         client,
+		scheme:         scheme,
+		serviceBuilder: knative.NewServiceBuilder(client, config),
+	}
+}
+
+func (r *ServiceReconciler) Reconcile(kfsvc *v1alpha2.KFService) error {
+	if err := r.reconcileDefault(kfsvc); err != nil {
+		return err
+	}
+
+	if err := r.reconcileCanary(kfsvc); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ServiceReconciler) reconcileDefault(kfsvc *v1alpha2.KFService) error {
+	for _, endpoint := range []constants.KFServiceEndpoint{constants.Predictor, constants.Transformer, constants.Explainer} {
+		if err := r.reconcileEndpoint(kfsvc, endpoint, false); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ServiceReconciler) reconcileCanary(kfsvc *v1alpha2.KFService) error {
+	for _, endpoint := range []constants.KFServiceEndpoint{constants.Predictor, constants.Transformer, constants.Explainer} {
+		if err := r.reconcileEndpoint(kfsvc, endpoint, true); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ServiceReconciler) reconcileEndpoint(kfsvc *v1alpha2.KFService, endpoint constants.KFServiceEndpoint, isCanary bool) error {
+	if isCanary {
+		if kfsvc.Spec.Canary == nil {
+			if err := r.finalizeService(kfsvc, endpoint); err != nil {
+				return err
+			}
+			kfsvc.Status.PropagateCanaryStatus(endpoint, nil)
+			return nil
+		}
+	}
+
+	service, err := r.serviceBuilder.CreateEndpointService(
+		kfsvc,
+		endpoint,
+		isCanary,
+	)
+
+	if err != nil {
+		return err
+	}
+
+	if service == nil {
+		return nil
+	}
+
+	status, err := r.reconcileService(kfsvc, service)
+	if err != nil {
+		return err
+	}
+	if isCanary {
+		kfsvc.Status.PropagateCanaryStatus(endpoint, status)
+	} else {
+		kfsvc.Status.PropagateDefaultStatus(endpoint, status)
+	}
+	return nil
+}
+
+func (r *ServiceReconciler) finalizeService(kfsvc *v1alpha2.KFService, endpoint constants.KFServiceEndpoint) error {
```

Comment:
If this is just for Canaries maybe we should change the name to reflect that?

Reply:
the function itself is generic, I'd suggest keep the name.